### ISSUE TITLE
Hashing algorithm not supported: "sha1"

### ISSUE THREAD

**Issue by harrypotter-magic** [chenxiaolong/avbroot#203]
While trying to unpack, patch and pack an system.img file I got the following error:

```
Error: Failed to update hash tree descriptor

Caused by:
    Hashing algorithm not supported: "sha1"
```

My commands were the following:

```bash
avbroot ota extract -a --input ota_update.zip # this is a patched zip, generated earlier with avbroot ota patch
avbroot avp unpack -i system.img
# edited raw.img, replaced a string there
avbroot avb pack -k avk.key --output system.img.patched
```

**Comment by chenxiaolong** [comment]
avbroot can unpack images with insecure hashing algorithms (like sha1), but it cannot pack them. However, you can just replace `sha1` with `sha256` inside `avb.toml`. No other changes are needed to switch to a secure hashing algorithm.

**Comment by harrypotter-magic** [comment]
I just looked at it and it looks like the [[header.descriptors]] KernelCmdline contains the string "sha1" and then the expected digest, and it doesn't look like avbroot modifies the hashing algorithm or the hash digest there. I wonder if the kernel validates it

**Comment by chenxiaolong** [comment]
> I just looked at it and it looks like the [[header.descriptors]] KernelCmdline contains the string "sha1" and then the expected digest, and it doesn't look like avbroot modifies the hashing algorithm or the hash digest there. I wonder if the kernel validates it

Interesting, I haven't seen that before. Can you post the full cmdline (and maybe a link to where I can download the entire OTA)?

**Comment by harrypotter-magic** [comment]
For example `https://download.lineageos.org/devices/taimen/builds`. I wonder if I can just remove the whole `KernelCmdline`, but I guess not. Maybe I can modify the cmdline to disable the hash check altogether.

**Comment by chenxiaolong** [comment]
Thanks, I've added support for automatically updating the kernel command line descriptor when packing an image in #205.

After I finish implementing the automatic `sha1` -> `sha256` promotion, I'll make a new avbroot release.

**Comment by chenxiaolong** [comment]
Hash algorithm promotion has been implemented in #206.

Both #205 and #206 have been released in avbroot 2.3.2.

---

I just saw this question that I forgot to answer.

> Maybe I can modify the cmdline to disable the hash check altogether.

I wouldn't recommend it, but you can do this by copying the `cmdline` value from the `KernelCmdline` descriptor containing `flags = 2` to the one containing `flags = 1`.

(The bootloader uses the `flags = 1` descriptor when verification is supposed to be enabled and the `flags = 2` descriptor when verification is disabled.)

**Comment by harrypotter-magic** [comment]
Thanks! it does get updated now, and gets promoted automatically.

I also noticed that the vbmeta.img conatins the kernel cmdline (or parts of it), is it ok? it contains the "sha1 8037329a60b5198af5159f65c830ac0d67696d12" part even after patching the image with `avbroot ota patch`. Is this part being validated as part of the avb?

**Comment by chenxiaolong** [comment]
Oof, that's a bug. The descriptor inside the `vbmeta` partition is supposed to be an exact copy. I'll look into it today.

**Comment by chenxiaolong** [comment]
The issue should be fixed with #208 (avbroot 2.3.3).
